internal/webhooks/patch_validation.go

```diff
@@ -183,20 +183,32 @@ func validateSelectors(selector clusterv1.PatchSelector, class *clusterv1.Cluste
 		}
 	}
 
-	if selector.MatchResources.MachineDeploymentClass != nil && len(selector.MatchResources.MachineDeploymentClass.Names) > 0 {
-		for _, name := range selector.MatchResources.MachineDeploymentClass.Names {
+	if selector.MatchResources.MachineDeploymentClass != nil {
+		if selector.MatchResources.MachineDeploymentClass.Names == nil {
```

Comment:
I think we should fail if the MachineDeploymentClass selector is defined (whether or not it's empty) if there are no MachineDeploymentClasses defined. i.e. if `len(class.Spec.Workers.MachineDeployments) == nil`

Reply:
But isn't that exactly the case we want to support? That `MachineDeploymentClass` is set to an empty object which means:
* selector.MatchResources.MachineDeploymentClass != nil  and
* selector.MatchResources.MachineDeploymentClass.Names == nil

Or am I missing something?

Reply:
Just based on https://github.com/kubernetes-sigs/cluster-api/pull/6930#issuecomment-1220702490 my idea was to reject `MachineDeploymentClass{Names=[]}`, but internally that looks exactly the same as `MachineDeploymentClass{}` no big issue though.

Reply:
What I meant here is we should fail if there is no MachineDeploymentClass defined in the ClusterClass at all i.e. the selector can't ever match anything.

Reply:
Ah sorry misread your comment. 

You mean PatchSelectorMatchMachineDeploymentClass.Names = [], right? (not `MachineDeploymentClass`)

That sounds fine to me, although I think this case already lands in l.213 given that neither if (l.187) nor  else if (l.196) will match

Reply:
Okay read the newer version of your comment. Looks like we already fail in that case today, but fine for me to improve the error message for that case.

Reply:
>I misunderstood the behaviour here this is working as expected.

Just to confirm: Looks like the for loop in l.188 will not trigger if there are no MachineDeploymentClasses are defined in the ClusterClass and it will fall to l.213 and return the correct error. 

Looks like it is working as expected. 👍